docs: added Vault documentation (#1224)
spectrocloud/librarium: content/docs/04.5-devx/05-app-profile/08-services/50-service-listings/4-vault.md

---
title: "Vault"
metaTitle: "Vault"
metaDescription: "Learn how to use Vault with Palette Dev Engine"
hideToC: false
type: "appTier"
category: ['security']
hiddenFromNav: false
fullWidth: false
logoUrl: "https://icon-library.com/images/padlock-icon-png/padlock-icon-png-29.jpg"
---

import Tabs from 'shared/components/ui/Tabs';
import WarningBox from 'shared/components/WarningBox';
import InfoBox from 'shared/components/InfoBox';

# Vault

> Palette Dev Engine (PDE) is leveraging Banzai Cloud's Bank-Vaults. Banzai Cloud Bank-Vaults is a wrapper for the official [Vault](https://www.vaultproject.io/) client. Vault is a tool that helps you manage and protect sensitive information, like passwords, API keys, and encryption keys, securely. The Bank-Vaults client enhances the offical Vault client by adding automatic token renewal, built-in Kubernetes support, and a dynamic database credential provider. 

Author: ```suggestion
Palette Dev Engine (PDE) is leveraging Banzai Cloud's Bank-Vaults. Banzai Cloud Bank-Vaults is a wrapper for the official [Vault](https://www.vaultproject.io/) client. Vault is a tool that helps you securely manage and protect sensitive information, like passwords, API keys, and encryption keys. The Bank-Vaults client enhances the official Vault client by adding automatic token renewal, built-in Kubernetes support, and a dynamic database credential provider. 
```

Vault keeps these secrets safe by locking them in a virtual "vault" and only allows authorized users to access them. Vault also tracks who has accessed which secrets and when making it easier to maintain security. You can use Vault to govern access to secrets, automate application delivery, and consume secrets programmatically.

Vault is deployed behind the scenes through the usage of the [Bank-Vaults Vault Operator Helm Chart](https://github.com/banzaicloud/bank-vaults/tree/main/charts/vault-operator).

<br />

<InfoBox>

Vault is deployed as a single container in the virtual cluster and the container is not tied to any particular node.

</InfoBox>

# Deploy Vault

Use the following steps to learn how to deploy Vault to your virtual cluster.

## Prerequisites

- A Spectro Cloud [account](https://www.spectrocloud.com/get-started/).

- A Virtual Cluster with at least the following minimum resources.
    - 4 CPU 
    - 6 GB of Memory

>     - 6 GB of Storage.

Author: Yeah that's a good point and they are both two different units of measurement.

|Description (optional)   | Description of the app profile. | 
|Tag (optional)           |  Assign tags to the app profile.|
 

4. Select the **Vault** service and start the configuration.
  

5. Provide the following information to the wizard:
  * **Name:** The application name.
  * **PersistentVolumeClaim Size (GiB):** Select the volume as per the storage volume available in the cluster group and virtual clusters. Ensure you do not exceed the maximum storage size for your virtual cluster. 

6. Save your changes.

7. Deploy the app profile to a Palette Virtual Cluster. Use the [Deploy a Virtual Cluster](/clusters/palette-virtual-clusters/deploy-virtual-cluster#deployavirtualcluster) guide for additional guidance or check out the [Deploy an Application using Palette Dev Engine](/devx/apps/deploy-app) tutorial.

## Validation

You can validate the Vault instance deployed successfully by using the following steps. 

1. Log in to [Palette](https://console.spectrocloud.com) and switch to **App Mode**.

2. Navigate to the left **Main Menu** and select **Apps**.

3. Select your application that contains Vault to view its details page.

> 4. Ensure the **Status** is *Deployed* and that the service *Vault* has a green dot right next to it.

Reviewer: ```suggestion
4. Ensure the **Status** is **Deployed** and that the **Vault** service has a green dot next to it.
```

5. Next, click on the **Virtual Cluster** link in the App details page.

> 6. Download the **kubeconfig** file by clicking on the url. 

Reviewer: ```suggestion
6. Click the URL to download the **kubeconfig**. 
```

> 7. Setup your local kubectl environment to use the kubeconfig file you downloaded. Review the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide for additional guidance.

Reviewer: ```suggestion
7. Set up your local kubectl environment to use the **kubeconfig** file you downloaded. Review the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide for additional guidance.
```

8. Export the following environment variables to prepare your environment to interact with Vault.

    <br />

    ```shell
    export VAULT_ADDR=https://127.0.0.1:8200 
    ```

    <br />

    ```shell
    export VAULT_SKIP_VERIFY=true
    ```

> 9. The next step is to configure port forwarding between your local workstation and the pod hosting Vault. Use the following commands to configure the port forward.

Reviewer: ```suggestion
9. Configure port forwarding between your local workstation and the pod hosting Vault. Use the following commands to configure the port forward.
```

    <br />

    ```shell
    VAULT_NAMESPACE=$(kubectl get pods --selector app.kubernetes.io/name=vault --all-namespaces --output jsonpath='{.items[0].metadata.namespace}') && \
    kubectl port-forward $(kubectl get pods --selector app.kubernetes.io/name=vault --all-namespaces --output jsonpath='{.items[0].metadata.name}') 8200:8200 --namespace $VAULT_NAMESPACE
    ```

    <br />

    ```shell
    kubectl port-forward $(kubectl get pods --selector app.kubernetes.io/name=vault --all-namespaces --output jsonpath='{.items[0].metadata.name}') 8200:8200 --namespace $VAULT_NAMESPACE
    ```

> 10. Open your browser and visit [https://localhost:8200/ui](https://localhost:8200/ui) to access the Vault UI. You will receive a warning due to the usage of a self-signed certificate but you can ignore this warning. 

Reviewer: This link is broken.

Author: It will work if the user follows the steps :) otherwise it will not work.

# Output Variables

The exposed output variables. Use these variables when connecting higher-level services with Vault:

| Parameter              | Output Variable                                                                     | Description                                     |
|------------------------|-------------------------------------------------------------------------------------|-------------------------------------------------|
| Vault Root Token | `{{.spectro.app.$appDeploymentName.<service-name>.VAULT_ROOT_TOKEN}}`              | The root token of the Vault instance. |
| Service Hostname       | `{{.spectro.app.$appDeploymentName.<service-name>.VAULTMSTR_SVC}}`      | The Kubernetes service hostname for the Vault service. |
| Service Port           | `{{.spectro.app.$appDeploymentName.<service-name>.VAULTMSTR_SVC_PORT}}` | The exposed port for the Vault service.              |
| Namespace           | `{{.spectro.app.$appDeploymentName.<service-name>.VAULTMSTR_SVC_NAMESPACE}}` | The Kubernetes namespace the Vault instance is deployed to. |

# Vault Credentials

> The Vault root token and the unseal keys are stored as a Kubernetes secret inside the virtual cluster. You can retrieve the Vault root token by using the following command.  <br /> <br />

Reviewer: ```suggestion
The Vault root token and the unseal keys are stored as a Kubernetes secret inside the virtual cluster. You can retrieve the Vault root token by following these steps.  <br /> <br />
```

1. Log in to [Palette](https://console.spectrocloud.com) and switch to **App Mode**.

2. Navigate to the left **Main Menu** and select **Clusters**.

3. Select the cluster that has Vault installed to view its details page.

4. Download the cluster **kubeconfig** file. 

> 5. Setup your local kubectl environment to use the kubeconfig you downloaded. Review the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide for additional guidance. 

Reviewer: ```suggestion
5. Set up your local kubectl environment to use the **kubeconfig** file you downloaded. Review the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide for additional guidance. 
```

> 6. You need get the Vault namespace and application name. Issue the following command to get the unique values.

Reviewer: ```suggestion
6. You need to get the Vault namespace and application name. Issue the following command to get the unique values.
```
